css/css-values/hypot-pow-sqrt-computed.html WPT crashes

rwlbuis opened the pull request:
#### 1643a89b579b0a44284018d1506f83090fd76a26
<pre>
css/css-values/hypot-pow-sqrt-computed.html WPT crashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=254392">https://bugs.webkit.org/show_bug.cgi?id=254392</a>

Reviewed by Tim Nguyen.

When simplifying hypot CSSCalcOperationNodes we may have only one child and
it may not be a CSSCalcPrimitiveValueNode, in that case do not simplify.

Add a similar check and tests for round, mod and rem.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed.html:
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::combineChildren):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:

Canonical link: <a href="https://commits.webkit.org/263345@main">https://commits.webkit.org/263345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c25f4b396686105d5dd1c8557883bfbf36244d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5737 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4721 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4327 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/4509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3845 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5732 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/2004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3821 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/6451 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5413 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4298 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3486 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3819 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1062 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4165 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nmoucht commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-computed-expected.txt:
```diff
@@ -10,12 +10,18 @@ PASS calc(1px * pow(2, 3)) should be used-value-equivalent to 8px
 PASS calc(1px * hypot(10000)) should be used-value-equivalent to 10000px
 PASS calc(2px * sqrt(100000000)) should be used-value-equivalent to 20000px
-PASS calc(3px * pow(200, 4)) should be used-value-equivalent to 33554428px
+PASS calc(3px * pow(20, 4)) should be used-value-equivalent to 480000px
```
Do we know why a few of these are failing?

rwlbuis replied:
Tim has a patch to fix that :) But I think it works better as a follow up patch.

nmoucht commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -650,26 +650,19 @@ void CSSCalcOperationNode::combineChildren()
     
     if (m_children.size() < 2) {
-        if (m_children.size() == 1 && isTrigNode()) {
+        if (isTrigNode() || isExpNode() || calcOperator() == CalcOperator::Sqrt) {
```
Should make a isSqrtNode() instead of directly checking the calc operator.

rwlbuis replied:
Done.

nmoucht commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -823,14 +810,14 @@ void CSSCalcOperationNode::combineChildren()
         m_children.append(WTFMove(newChild));
     }
-    if (isRoundOperation()) {
+    if (isRoundOperation() && canCombineAllChildren()) {
```
Is there a reason this check isn't necessary for the rest of the calc functions?

rwlbuis replied:
Not all calc functions end up in combineChildren() I think and a lot of calc functions only deal with primitive values.

nmoucht commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -650,26 +650,19 @@ void CSSCalcOperationNode::combineChildren()
             m_children.append(WTFMove(newChild));
         }
-        
-        if (isExpNode()) {
```
I think we can do a similar refactoring for stepped node and round node below.

rwlbuis replied:
Done.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed.html:
Can you export those changes to WPT?

rwlbuis replied:
Sure, I have a PR here https://github.com/web-platform-tests/wpt/pull/39679.

rwlbuis committed:
css/css-values/hypot-pow-sqrt-computed.html WPT crashes
https://bugs.webkit.org/show_bug.cgi?id=254392

Reviewed by Tim Nguyen.

When simplifying hypot CSSCalcOperationNodes we may have only one child and
it may not be a CSSCalcPrimitiveValueNode, in that case do not simplify.

Add a similar check and tests for round, mod and rem.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed.html:
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::combineChildren):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:

Canonical link: https://commits.webkit.org/263345@main
